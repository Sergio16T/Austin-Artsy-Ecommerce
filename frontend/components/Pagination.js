import React, { Component } from 'react';
import PaginationStyles from './styles/PaginationStyles'; 
import { Query } from 'react-apollo'; 
import Head from "next/head"; 
import Link from "next/link"; 
import styled from 'styled-components'; 
import gql from 'graphql-tag'; 

const perPage = 4; 

const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY{
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`;


const PaginationContainer = styled.div`
    padding-top: 4rem; 
    width: 100%; 
    display: flex; 
    justify-content: center; 
`;
const Placeholder = styled.div`
    height: 45px;
    margin-top: 24px; 
    margin-bottom: 24px; 
`; 
const Pagination = props => {
    return (
        <PaginationContainer>
            <Query query={PAGINATION_QUERY}>
                {({data, error, loading}) => {
                    if(loading) return <Placeholder/>; 
                    if(error) return <p>{error.message}</p>
                    const count = data.itemsConnection.aggregate.count; 
                    const pages = Math.ceil(count / perPage); 
                    return (
                    <PaginationStyles data-testid="pagination">
                        <Head>
                            <title>Austin's List page {props.page} of {pages}</title>
                        </Head>
                        <Link                      
                        href={{
                            pathname: "/items",
                            query: {page: props.page -1}
                        }}
                        >
                        <a aria-disabled={props.page <= 1} id="prev"> &#8592; Prev</a>
                        </Link>
                        <p data-test="pagecount">Page {props.page} of {pages}</p>
                        <Link                      
                        href={{
                            pathname:"/items",
                            query: {page: props.page + 1}
                        }}

                        >
                        <a aria-disabled={props.page >= pages} id="next">Next &#8594;</a>
                        </Link>
                    </PaginationStyles>
                    )}}
            </Query> 
        </PaginationContainer>
    );
};


export default Pagination;
export { PAGINATION_QUERY }; 