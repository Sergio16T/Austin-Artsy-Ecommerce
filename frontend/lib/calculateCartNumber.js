export default function(cart) {
    return cart.reduce((startValue, element)=> {
        return startValue + element.quantity; 
    }, 0); 
}