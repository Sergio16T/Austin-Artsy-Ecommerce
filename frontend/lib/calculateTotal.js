export default function(cart) {
    return cart.reduce((startValue, element) => {
    if(!element.item) return startValue; 
    return startValue + (element.quantity * element.item.price)}, 0); 
    
}