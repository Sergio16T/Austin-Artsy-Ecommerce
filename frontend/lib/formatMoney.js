export default function(amount) {
    let price = amount / 100; 
    return new Intl.NumberFormat('en', {
        style: "currency",
        currency: "USD",
    }).format(price);
}

/* when receiving money from form make sure it always has two floating point numbers and then just multiply it by 100
to store it in graphql as cents in int type then when I'm displaying money multiply it by 100.
that way I can format it for use later.. so it's always going to be handled in cents 
165.33 * 100
16533
165 * 100
16500
16500/100
165
16533/100
165.33
 */ 
