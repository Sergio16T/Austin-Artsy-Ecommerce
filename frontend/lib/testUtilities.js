import casual from 'casual'; 
//.seed() is to provide ability to use a specific seed in order to get a repeatable random sequence for our testing purposes
casual.seed(111); 

const mockItem = () => ({
    // __typename: 'Item',
    id: '123',
    price: 5000,
    user: null,
    image: 'dog.jpg',
    title: 'dogs are the best',
    description: 'dogs description',
    largeImage: 'dog-large.jpg',
}); 

const mockUserWithItemsInCart = () =>  ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ["ADMIN"], 
    cart: [{
        id: 'abc', 
        quantity: 2, 
        item: {
            id: 'dgh', 
            price: 15000, 
            image: 'www.dog.image.com',
            title: "puppy"
        }
        },
    ]
}); 
const mockUser = () => ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ['ADMIN'], 
    cart: [],
    orders: []
}); 
const mockOrderItem = () => ({
    __typename: 'OrderItem',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
  });
  
  const mockOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    items: [mockOrderItem(), mockOrderItem()],
    createdAt: '2018-04 - 06T19: 24: 16.000Z',
    user: mockUser(),
  });
  
  const mockCartItem = overrides => ({
    __typename: 'CartItem',
    id: 'omg123',
    quantity: 3,
    item: mockItem(),
    user: mockUser(),
    ...overrides,
  });
//mock LocalStorage 
class MockLocalStorage {
    constructor(){
        this.store={}; 
    }

    clear() {
        this.store = {}; 
    }
    getItem(key) {
        return this.store[key] || null; 
    }
    setItem(key, value) {
        this.store[key] = value.toString(); 
    }
    removeItem(key) {
        delete this.store[key]; 
    }
}

export { 
    mockItem, 
    MockLocalStorage, 
    mockUser, 
    mockCartItem, 
    mockOrderItem, 
    mockOrder }; 