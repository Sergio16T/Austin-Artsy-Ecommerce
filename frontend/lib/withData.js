import withApollo from 'next-with-apollo'; 
import ApolloClient from 'apollo-boost'; 
import { endpoint } from '../config'; 

function createClient({ headers}) {
    return new ApolloClient({
        uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint, 
        // on each request pass the headers and setContext change credentials to 'include' for whether the user agent should send cookies from other domains in case of CORS
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include', // logged in cookies present ? include.  (default value is same-origin) 
                },
                headers, 
            }); 
        },
    }); 
}

export default withApollo(createClient); 