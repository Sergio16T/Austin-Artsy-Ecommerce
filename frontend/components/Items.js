import React, { Component } from 'react';
import { Query } from 'react-apollo'; 
import gql from 'graphql-tag';
import Item from './Item';
import { ItemStyles, StyledItemsWithPagination, StyledItemContainer } from './styles/ItemStyles'; 
import Pagination from './Pagination'; 

const ALL_ITEMS_QUERY = gql`
    query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
        items(skip: $skip, first: $first, orderBy: createdAt_ASC) {
            id
            title
            price
            description
            image 
            largeImage
        }
    }
`; 


class Items extends Component {
    render() {
        return (
            <ItemStyles>
                <Query query={ALL_ITEMS_QUERY}
                variables={{skip: this.props.page * 4 - 4}}
                >
                    {({ data, error, loading }) => { 
                        if (loading) return null; 
                        if (error) return null; 
                        // console.log('payload', data);
                        return (
                            <StyledItemsWithPagination>
                                <Pagination page={this.props.page}/>
                                        <StyledItemContainer>    
                                            {data.items.map(item => (
                                                <Item 
                                                key={item.id} 
                                                item={item} 
                                                page={this.props.page}
                                                toggleCart={this.props.toggleCart}
                                                />           
                                            ))}
                                        </StyledItemContainer>  
                                <Pagination page={this.props.page}/>
                            </StyledItemsWithPagination>
                        )
                    }}
                </Query>
            </ItemStyles>
        );
    }
}

export default Items;
export { ALL_ITEMS_QUERY }
