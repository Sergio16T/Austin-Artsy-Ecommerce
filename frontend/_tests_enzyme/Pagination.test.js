import { mount } from 'enzyme'; 
import { shallowToJson } from 'enzyme-to-json'; 
import Items, { ALL_ITEMS_QUERY } from '../components/Items'; 
import Pagination, { PAGINATION_QUERY } from '../components/Pagination'; 
import { MockedProvider } from '@apollo/react-testing'; 
import { mockUser, mockItem } from '../lib/testUtilities'; 
import { act } from 'react-dom/test-utils'; 
import wait from 'waait';  


const mockPagination = (length) => {
    return [{
        request: { query: PAGINATION_QUERY }, 
        result: { 
            data: {
                itemsConnection: {
                    __typename: 'aggregate',
                    aggregate: {
                        __typename: 'count', 
                        count: length
                    }
                },
            },
        },
    }]
}

describe('<Pagination/>', () => {
    it('mounts with correct number of pages', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mockPagination(8)}>
                <Pagination page={1}/>
            </MockedProvider>
        ); 
        await wait(); 
        wrapper.update(); 
        // console.log(wrapper.debug()); 
        expect(wrapper.find('p[data-test="pagecount"]').text()).toEqual('Page 1 of 2'); 
        const pagination = wrapper.find('div[data-test="pagination"]'); 
        console.log(pagination.debug()); 
        expect(shallowToJson(pagination)).toMatchSnapshot(); 
    });
    it('disables prev button on first page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mockPagination(8)}>
                <Pagination page={1}/>
            </MockedProvider>
        ); 
        await wait(); 
        wrapper.update(); 
        const prevBtn = wrapper.find('a#prev'); 
        const nextBtn = wrapper.find('a#next'); 
        expect(prevBtn.prop('aria-disabled')).toEqual(true); 
        expect(nextBtn.prop('aria-disabled')).toEqual(false); 
        // console.log(prevBtn.debug()); 
    }); 
    it('disables next button on last page', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mockPagination(22)}>
                <Pagination page={6}/>
            </MockedProvider>
        ); 
        await wait(); 
        wrapper.update(); 
        const prevBtn = wrapper.find('a#prev'); 
        const nextBtn = wrapper.find('a#next'); 
        expect(prevBtn.prop('aria-disabled')).toEqual(false); 
        expect(nextBtn.prop('aria-disabled')).toEqual(true); 
        // console.log(prevBtn.debug()); 
    }); 
    it('enables next and prev button on page in the middle', async () => {
        const wrapper = mount(
            <MockedProvider mocks={mockPagination(22)}>
                <Pagination page={4}/>
            </MockedProvider>
        ); 
        await wait(); 
        wrapper.update(); 
        const prevBtn = wrapper.find('a#prev'); 
        const nextBtn = wrapper.find('a#next'); 
        expect(prevBtn.prop('aria-disabled')).toEqual(false); 
        expect(nextBtn.prop('aria-disabled')).toEqual(false); 
        // console.log(prevBtn.debug()); 
    }); 
});