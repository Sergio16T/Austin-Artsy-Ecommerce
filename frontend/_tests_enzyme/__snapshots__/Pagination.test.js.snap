// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> mounts with correct number of pages 1`] = `
<div
  className="PaginationStyles-cr7hxd-0 hzzTYo"
  data-test="pagination"
>
  <Head>
    <Component
      handleStateChange={null}
      inAmpMode={false}
      reduceComponentsToState={[Function]}
    />
  </Head>
  <Link
    href={
      Object {
        "pathname": "/items",
        "query": Object {
          "page": 0,
        },
      }
    }
  >
    <a
      aria-disabled={true}
      href="/items?page=0"
      id="prev"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
       ← Prev
    </a>
  </Link>
  <p
    data-test="pagecount"
  >
    Page 
    1
     of 
    2
  </p>
  <Link
    href={
      Object {
        "pathname": "/items",
        "query": Object {
          "page": 2,
        },
      }
    }
  >
    <a
      aria-disabled={false}
      href="/items?page=2"
      id="next"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Next →
    </a>
  </Link>
</div>
`;
